Python/gc_free_threading.c

```diff
@@ -1025,6 +1025,11 @@ gc_collect_internal(PyInterpreterState *interp, struct collection_state *state)
     merge_all_queued_objects(interp, state);
     process_delayed_frees(interp);
 
+    // Free all outstanding objects.  During finalization we could collect
+    // types before the underlying objects so its important that we not
+    // traverse any delay-freed objects.
+    _PyMem_ProcessAllDelayed(interp);
```

Comment:
I think think this is the same thing as `process_delayed_frees` above?

The code comment makes it sound like there's a correctness concern if we don't free all delayed-free objects. I'm worried that relying on that assumption too much makes the code fragile.

If I understand correctly, the problematic situation occurs when:
1) We delay-free a GC object
2) The object's destructor does not untrack it (if it's untracked then it's not a concern)

I think we should be sure to untrack the object in `PyObject_GC_Del` immediately before calling `_PyObject_FreeDelayed()`. The default GC already has [similar behavior](https://github.com/python/cpython/blob/58c7919d05a360f3437a204960cddbeb78a71dce/Python/gc.c#L1899), but it got removed in the free-threaded GC because untracking before freeing didn't seem necessary previously.

Reply:
The actual issue is apparently only happens in debug builds (I hadn't realized this before), but it's that we're calling `PyObject_IS_GC` during finalization of the runtime, and the type has been freed.  The object isn't tracked but the memory is still going to be visible during the walk.  The crashing stack is:

```
#0  PyObject_IS_GC (obj=obj@entry=0x200005b05d0) at Python/gc_free_threading.c:1569
#1  0x0000000000686449 in op_from_block (block=<optimized out>, arg=<optimized out>, include_frozen=include_frozen@entry=false) at Python/gc_free_threading.c:219
#2  0x0000000000686727 in update_refs (heap=<optimized out>, area=<optimized out>, block=<optimized out>, block_size=<optimized out>, args=<optimized out>) at Python/gc_free_threading.c:347
#3  0x0000000000542635 in _mi_heap_area_visit_blocks (area=area@entry=0x7fffffffd1c0, page=<optimized out>, visitor=0x686710 <update_refs>, arg=0x7fffffffd3c0) at Objects/mimalloc/heap.c:624
#4  0x00000000005426aa in mi_heap_area_visitor (heap=heap@entry=0xa21020 <_PyRuntime+284832>, xarea=xarea@entry=0x7fffffffd1c0, arg=arg@entry=0x7fffffffd290) at Objects/mimalloc/heap.c:675
#5  0x00000000005421c1 in mi_heap_visit_areas_page (heap=0xa21020 <_PyRuntime+284832>, pq=<optimized out>, page=<optimized out>, vfun=0x54266a <mi_heap_area_visitor>, arg=0x7fffffffd290) at Objects/mimalloc/heap.c:655
#6  0x0000000000539824 in mi_heap_visit_pages (heap=heap@entry=0xa21020 <_PyRuntime+284832>, fn=fn@entry=0x54218f <mi_heap_visit_areas_page>, arg1=arg1@entry=0x54266a <mi_heap_area_visitor>, arg2=arg2@entry=0x7fffffffd290)
    at Objects/mimalloc/heap.c:46
#7  0x0000000000539887 in mi_heap_visit_areas (heap=heap@entry=0xa21020 <_PyRuntime+284832>, visitor=visitor@entry=0x54266a <mi_heap_area_visitor>, arg=arg@entry=0x7fffffffd290) at Objects/mimalloc/heap.c:661
#8  0x000000000054613b in mi_heap_visit_blocks (heap=heap@entry=0xa21020 <_PyRuntime+284832>, visit_blocks=visit_blocks@entry=true, visitor=visitor@entry=0x686710 <update_refs>, arg=arg@entry=0x7fffffffd3c0) at Objects/mimalloc/heap.c:686
#9  0x00000000006843dd in gc_visit_heaps_lock_held (interp=interp@entry=0x9f8840 <_PyRuntime+118976>, visitor=visitor@entry=0x686710 <update_refs>, arg=arg@entry=0x7fffffffd3c0) at Python/gc_free_threading.c:253
#10 0x000000000068449d in gc_visit_heaps (interp=interp@entry=0x9f8840 <_PyRuntime+118976>, visitor=visitor@entry=0x686710 <update_refs>, arg=arg@entry=0x7fffffffd3c0) at Python/gc_free_threading.c:287
#11 0x00000000006847fd in deduce_unreachable_heap (interp=interp@entry=0x9f8840 <_PyRuntime+118976>, state=state@entry=0x7fffffffd3c0) at Python/gc_free_threading.c:514
#12 0x000000000068565d in gc_collect_internal (interp=interp@entry=0x9f8840 <_PyRuntime+118976>, state=state@entry=0x7fffffffd3c0) at Python/gc_free_threading.c:1016
#13 0x0000000000687274 in gc_collect_main (tstate=tstate@entry=0xa1eac8 <_PyRuntime+275272>, generation=generation@entry=2, reason=reason@entry=_Py_GC_REASON_SHUTDOWN) at Python/gc_free_threading.c:1138
#14 0x00000000006874ba in _PyGC_CollectNoFail (tstate=tstate@entry=0xa1eac8 <_PyRuntime+275272>) at Python/gc_free_threading.c:1459
#15 0x00000000006c09b7 in finalize_modules (tstate=tstate@entry=0xa1eac8 <_PyRuntime+275272>) at Python/pylifecycle.c:1711
#16 0x00000000006cab91 in Py_FinalizeEx () at Python/pylifecycle.c:1973
#17 0x00000000006fd162 in Py_RunMain () at Modules/main.c:709
#18 0x00000000006fd1b1 in pymain_main (args=args@entry=0x7fffffffd520) at Modules/main.c:737
#19 0x00000000006fd226 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>) at Modules/main.c:761
#20 0x000000000041d84f in main (argc=<optimized out>, argv=<optimized out>) at ./Programs/python.c:15
```

I can add the untracking behavior too though.